File: Sources/AndroidBackend/Kotlin/CustomWebView.kt
package dev.swiftcrossui.androidbackend

import android.app.Activity
import android.webkit.WebResourceRequest
import android.webkit.WebView
import android.webkit.WebViewClient

class CustomWebView(activity: Activity): WebView(activity) {
    var onNavigate: SwiftAction? = null

    var loadingUrl: String? = null
        private set
    
    init {
        webViewClient = object : WebViewClient() {
            override fun shouldOverrideUrlLoading(
                webView: WebView,
                request: WebResourceRequest
            ): Boolean {
                loadingUrl = request.url?.toString()
                onNavigate?.call()
                return false
            }
        }

        settings.javaScriptEnabled = true
    }
    
    override fun loadUrl(url: String) {
        loadingUrl = url
        super.loadUrl(url)
    }
}

File: Sources/AndroidBackend/CustomWebView.swift
import SwiftJava
import AndroidKit

@JavaClass(
    "dev.swiftcrossui.androidbackend.CustomWebView",
    extends: AndroidKit.WebView.self
)
class CustomWebView: JavaObject {
    @JavaMethod
    @_nonoverride convenience init(
        _ activity: Activity?,
        environment: JNIEnvironment? = nil
    )

    @JavaMethod
    func setOnNavigate(_ action: SwiftAction?)

    @JavaMethod
    func loadUrl(_ url: String)
    
    @JavaMethod
    func getLoadingUrl() -> JavaString?
}
